aptos-move/framework/aptos-token/sources/bid.move

```diff
@@ -0,0 +1,343 @@
+        assert!(table::contains(bid_records, bid_id), EBID_NOT_EXIST);
+        let bid = table::borrow(bid_records, bid_id);
+
+        let coin_owner = bid.bidder;
+        let token_owner = listing::get_listing_owner(entry);
+        // validate offerred amount and price
+        let listed_amount =  listing::get_listing_token_amount(entry);
+        let min_total = listing::get_listing_min_price(entry) * listed_amount;
+        assert!(coin::value(&bid.coin) >= min_total, ENO_SUFFICIENT_FUND);
+        // validate expiration time
+        let now = timestamp::now_seconds();
+        // listing should expire after auction ends
+        assert!(listing::get_listing_expiration(entry) >= now, ELISTING_EXPIRED);
+        assert!(listing::get_listing_start(entry) <= now, ELISTING_NOT_STARTED);
+        //listing_id matches
+        assert!(listing::get_listing_id(entry) == bid.listing_id, ELISTING_ID_NOT_MATCH);
+
+        // transfer coin and token
+        let token = token::withdraw_with_event_internal(
+            token_owner,
+            listing::get_listing_token_id(entry),
+            listed_amount
+        );
+
+        token::direct_deposit(coin_owner, token);
+
+        let bid_mut = table::remove(bid_records, bid_id);
+        let coins = clear_bid(bid_mut);
+
+        // deduct royalty fee from the transactions
+        let royalty = token::get_royalty(listing::get_listing_token_id(entry));
+        let royalty_payee = token::get_royalty_payee(&royalty);
+        let royalty_coin = deduct_fee(
+            &mut coins,
+            token::get_royalty_numerator(&royalty),
+            token::get_royalty_denominator(&royalty)
+        );
+        coin::deposit(royalty_payee, royalty_coin);
+
+        // deduct marketplace fee
+        let market_fee = deduct_fee(&mut coins, fee_numerator, fee_denominator);
+        coin::deposit(market_address, market_fee);
+
+        // give the remaining to the seller
+        coin::deposit(token_owner, coins);
+    }
+
+    /// destruct the bid struct and extract coins
+    fun clear_bid<CoinType>(bid: Bid<CoinType>): Coin<CoinType> {
+        let Bid {
+            bidder: _,
+            coin,
+            offer_price: _,
+            listing_id: _,
+        } = bid;
+        coin
+    }
+
+    public fun get_bid_info<CoinType>(
+        bidder_address: address,
+        bid_id_creation_number: u64
+    ): (u64, ID) acquires BidRecords {
+        let bid_id = guid::create_id(bidder_address, bid_id_creation_number);
+
+        let bid_records = &mut borrow_global_mut<BidRecords<CoinType>>(bidder_address).records;
+        assert!(table::contains(bid_records, bid_id), EBID_NOT_EXIST);
+
+        let bid = table::borrow(bid_records, bid_id);
+        ( bid.offer_price, bid.listing_id )
+    }
+
+    /// internal function for assigned a global unique id for a listing
+    fun create_bid_id(owner: &signer): ID {
+        let gid = account::create_guid(owner);
+        guid::id(&gid)
+    }
```

Comment:
There's no point in using a guid here, we can just as well use a u64 counter. If we aren't sharing the GUID across users, then address portion is irrelevant and the u64 is globally unique for that user (bid resource + user account + u64).

Reply:
a global counter shared by multiple users will limit parallelism.